File: src/lib/c++/gmp_lib/src/GMP/GMP_Update.cpp
#include <gmp_lib/GMP/GMP_Update.h>

namespace as64_
{

namespace gmp_
{

  GMP_Update::GMP_Update(gmp_::GMP *gmp)
  {
      this->gmp = gmp;

      this->initSigmaw();
      this->rv = 1.0;

      this->recursiveUpdate(false);
      this->syncUpdate(true);
  }

  void GMP_Update::syncUpdate(bool set)
  {
    this->sync_up = set;
  }

  void GMP_Update::recursiveUpdate(bool set)
  {
      this->recursive_up = set;
  }

  void GMP_Update::initSigmaw()
  {
      unsigned N_kernels = gmp->numOfKernels();
      // arma::mat S = arma::mat().zeros(N_kernels,N_kernels);
      // for (int i=0; i<N_kernels; i++)
      // {
      //     for (int j=i+1; j<N_kernels; j++)
      //     {
      //         S(i,j) = std::exp(-0.2 * std::abs(i-j));
      //     }
      // }
      // S = S + S.t() + arma::mat().eye(N_kernels,N_kernels);
      // this->Sigma_w = S;
      this->Sigma_w = arma::mat().eye(N_kernels, N_kernels);
  }

  void GMP_Update::initExpSigmaw(double decay_rate)
  {      
    unsigned N_kernels = this->gmp->numOfKernels();
    arma::mat S = arma::mat().zeros(N_kernels, N_kernels);
    for (int i=0; i<N_kernels; i++)
    {
      for (int j=i+1; j<N_kernels; j++) S(i,j) = std::exp(-decay_rate * std::abs(i-j));
    }
    this->Sigma_w = S + S.t() + arma::mat().eye(N_kernels, N_kernels); 
  }

  void GMP_Update::initSigmaWfromPosMsr(const arma::rowvec &x_data)
  {
    unsigned n_data = x_data.size();
    arma::mat H(gmp->numOfKernels(), n_data);
    for (int j=0; j<n_data; j++) H.col(j) = gmp->regressVec(x_data(j));

    this->Sigma_w = arma::inv(H*H.t());
  }

  void GMP_Update::initSigmaWfromVelMsr(const arma::rowvec &x_data, double Tf)
  {
    unsigned n_data = x_data.size();
    arma::mat H(gmp->numOfKernels(), n_data);
    for (int j=0; j<n_data; j++) H.col(j) = gmp->regressVecDot(x_data(j), 1.0/Tf);

    this->Sigma_w = arma::inv(H*H.t());
  }

  void GMP_Update::initSigmaWfromAccelMsr(const arma::rowvec &x_data, double Tf)
  {
    unsigned n_data = x_data.size();
    arma::mat H(gmp->numOfKernels(), n_data);
    for (int j=0; j<n_data; j++) H.col(j) = gmp->regressVecDDot(x_data(j), 1.0/Tf, 0.0);

    this->Sigma_w = arma::inv(H*H.t());
  }
  
  void GMP_Update::setSigmaW(const arma::mat &Sw)
  {
    this->Sigma_w = Sw;
  }

  arma::mat GMP_Update::getSigmaW() const
  {
    return this->Sigma_w;
  }

  void GMP_Update::setMsrNoiseVar(double rv)
  {
    this->rv = rv;
  }

  // ==================================================
  // ===============   Online update  =================
  // ==================================================

  void GMP_Update::updateNow()
  {
    updateWeights(batch_s, batch_Z, batch_type, batch_r_n);
    clearBatch();
  }

  void GMP_Update::clearBatch()
  {
    batch_s.clear();
    batch_Z.clear();
    batch_type.clear();
    batch_r_n.clear();
  }

  void GMP_Update::updatePos(double x, const arma::vec &y, double r_n)
  {
    if (r_n < 0) r_n=this->rv;
    
    batch_s.push_back(gmp_::Phase(x,0,0));
    batch_Z = arma::join_horiz(batch_Z, y);
    batch_type.push_back(gmp_::UPDATE_TYPE::POS);
    batch_r_n = arma::join_horiz(batch_r_n, arma::rowvec({r_n}));

    if (sync_up) updateNow();
  }

  void GMP_Update::updateVel(double x, double x_dot, const arma::vec &y_dot, double r_n)
  {
    if (r_n < 0) r_n=this->rv;
    
    batch_s.push_back(gmp_::Phase(x,x_dot,0));
    batch_Z = arma::join_horiz(batch_Z, y_dot);
    batch_type.push_back(gmp_::UPDATE_TYPE::VEL);
    batch_r_n = arma::join_horiz(batch_r_n, arma::rowvec({r_n}));

    if (sync_up) updateNow();
  }

  void GMP_Update::updateAccel(double x, double x_dot, double x_ddot, const arma::vec &y_ddot, double r_n)
  {
    if (r_n < 0) r_n=this->rv;
    
    batch_s.push_back(gmp_::Phase(x,x_dot,x_ddot));
    batch_Z = arma::join_horiz(batch_Z, y_ddot);
    batch_type.push_back(gmp_::UPDATE_TYPE::ACCEL);
    batch_r_n = arma::join_horiz(batch_r_n, arma::rowvec({r_n}));

    if (sync_up) updateNow();
  }

  void GMP_Update::updateWeights(const std::vector<gmp_::Phase> &s, arma::mat Z, const std::vector<gmp_::UPDATE_TYPE> &type, arma::rowvec r_n)
  {
    if (r_n.is_empty()) r_n = {this->rv};

    unsigned n = Z.n_cols;
    unsigned n_ker = gmp->numOfKernels();
    arma::mat sc = gmp->getScaling();
    arma::mat inv_sc = gmp->getInvScaling();

    if (r_n.size() == 1) r_n = r_n(0)*arma::rowvec().ones(n);
    arma::mat Rn = arma::diagmat(r_n);

    arma::mat H(n_ker, n);

    for (int j=0; j<n; j++)
    {
        if (type[j] == gmp_::UPDATE_TYPE::POS)
        {
            H.col(j) = gmp->regressVec(s[j].x);
            Z.col(j) = Z.col(j) - gmp->Y0 + sc*gmp->Y0d;
        }
        else if (type[j] == gmp_::UPDATE_TYPE::VEL)
            H.col(j) = gmp->regressVecDot(s[j].x, s[j].x_dot);
        else // (type == gmp_::UPDATE_TYPE::ACCEL)
            H.col(j) = gmp->regressVecDDot(s[j].x, s[j].x_dot, s[j].x_ddot);
        //Z.col(j) = Z.col(j)/sc;
    }
    Z = inv_sc*Z;

    arma::mat C = this->Sigma_w*H;
    arma::mat B = arma::inv_sympd(H.t()*this->Sigma_w*H + Rn) * C.t();
    gmp->W = gmp->W + (Z - gmp->W*H)*B;

    if (this->recursive_up) this->Sigma_w = this->Sigma_w - C*B;
  }

} // namespace gmp_

} // namespace as64_

File: src/lib/matlab/gmp_lib/GMP/GMP_Update.m
%% N-DoF GMP Update class
%

classdef GMP_Update < matlab.mixin.Copyable

    methods (Access = public)
        
        %% GMP constructor.
        %  @param[in] gmp: n_DoF dmp.
        function this = GMP_Update(gmp)
                
%             if ( ~ strcmpi(class(gmp),'GMP') )
%                 error(['[GMP_Update::GMP_Update]: argument of type ''GMP'' was expected, but a type ''' class(gmp) ''' was provided instead']);
%             end
            
            this.gmp = gmp;
            
            this.initSigmaw();
            this.rv = 1;
            
            this.recursiveUpdate(false);
            this.syncUpdate(true);

        end

        function initSigmaw(this)
            
            N_kernels = this.gmp.numOfKernels();
            this.Sigma_w = eye(N_kernels, N_kernels);
            
        end
        
        function initExpSigmaw(this, decay_rate)
            
            if (nargin < 2), decay_rate = 0.01; end
            
            N_kernels = this.gmp.numOfKernels();
            S = zeros(N_kernels,N_kernels);
            for i=1:N_kernels
                for j=i+1:N_kernels
                    S(i,j) = exp(-decay_rate * abs(i-j));
                end
            end
            this.Sigma_w = S + S' + eye(N_kernels,N_kernels); 
  
        end
        
        function initSigmaWfromMsr(this, x_data)
              
            n_data = length(x_data);
            H = zeros(this.gmp.numOfKernels(), n_data);
            for j=1:n_data, H(:,j) = this.gmp.regressVec(x_data(j)); end
            
            this.Sigma_w = inv(H*H');
            
        end
        
        function initSigmaWfromVelMsr(this, x_data, tau)
            
            n_data = length(x_data);
            H = zeros(this.gmp.numOfKernels(), n_data);
            for j=1:n_data, H(:,j) = this.gmp.regressVecDot(x_data(j), 1/tau); end
            
            this.Sigma_w = inv(H*H' + 1e-10*eye(this.gmp.numOfKernels()));
            
        end
        
        function initSigmaWfromAccelMsr(this, x_data, tau)
            
            n_data = length(x_data);
            H = zeros(this.gmp.numOfKernels(), n_data);
            for j=1:n_data, H(:,j) = this.gmp.regressVecDDot(x_data(j), 1/tau, 0); end
            
            this.Sigma_w = inv(H*H' + 1e-10*eye(this.gmp.numOfKernels()));
            
        end
        
        function recursiveUpdate(this, set)
           
            this.recursive_up = set;
            
        end
        
        function syncUpdate(this, set)
           
            this.sync_up = set;
            
        end
        
        function setSigmaW(this, Sw)
        
            this.Sigma_w = Sw;
            
        end
        
        function Sw = getSigmaW(this)
           
            Sw = this.Sigma_w;
            
        end
        
        function setMsrNoiseVar(this, rv)
           
            this.rv = rv;
            
        end
        
        %% ==================================================
        %% ===============   Online update  =================
        %% ==================================================
        
        function updateNow(this)
            
            this.updateWeights(this.batch_s, this.batch_Z, this.batch_type, this.batch_r_n);
            this.clearBatch();
    
        end

        function updatePos(this, x, y, r_n)
            
            if (nargin < 4), r_n=this.rv; end
            
            this.batch_s = [this.batch_s, GMP_phase(x,0,0)];
            this.batch_Z = [this.batch_Z, y];
            this.batch_type = [this.batch_type, GMP_UpdateType.POS];
            this.batch_r_n = [this.batch_r_n, r_n];

            if (this.sync_up), this.updateNow(); end
            
        end
        
        function updateVel(this, x, x_dot, y_dot, r_n)
            
            if (nargin < 5), r_n=this.rv; end

            this.batch_s = [this.batch_s, GMP_phase(x,x_dot,0)];
            this.batch_Z = [this.batch_Z, y_dot];
            this.batch_type = [this.batch_type, GMP_UpdateType.VEL];
            this.batch_r_n = [this.batch_r_n, r_n];

            if (this.sync_up), this.updateNow(); end
            
        end
        
        function updateAccel(this, x, x_dot, x_ddot, y_ddot, r_n)
            
            if (nargin < 6), r_n=this.rv; end

            this.batch_s = [this.batch_s, GMP_phase(x,x_dot,x_ddot)];
            this.batch_Z = [this.batch_Z, y_ddot];
            this.batch_type = [this.batch_type, GMP_UpdateType.ACCEL];
            this.batch_r_n = [this.batch_r_n, r_n];

            if (this.sync_up), this.updateNow(); end
            
        end
        
        
        %% Updates the weights based on 'n' measurements.
        %  @param[in] s: 1 x n vector of type GMP_phase, where the j-th entry is the phase for the j-th measurement.
        %  @param[in] Z: n_dof x n matrix, where the j-th column is the j-th measurement.
        %  @param[in] type: 1 x n vector of type GMP_UpdateType, where the j-th entry has the update type for the j-th measurement.
        %  @param[in] r_n: 1 x n vector, where the j-th entry is the noise variance for the j-th measurement (optional, default=this.rv).
        function updateWeights(this, s, Z, type, r_n)
 
            if (nargin < 5), r_n = this.rv; end
                
            n = size(Z,2);
            n_ker = this.gmp.numOfKernels();
            sc = this.gmp.getScaling();
            inv_sc = this.gmp.getInvScaling();
            
            if (length(r_n) == 1), r_n = repmat(r_n(1), 1,n); end
            Rn = diag(r_n);
 
            H = zeros(n_ker, n);
            
            for j=1:n
                if (type(j) == GMP_UpdateType.POS)
                    H(:,j) = this.gmp.regressVec(s(j).x);
                    Z(:,j) = Z(:,j) - this.gmp.Y0 + sc*this.gmp.Y0d;
                elseif (type(j) == GMP_UpdateType.VEL)
                    H(:,j) = this.gmp.regressVecDot(s(j).x, s(j).x_dot);
                else % (type == GMP_UpdateType.ACCEL)
                    H(:,j) = this.gmp.regressVecDDot(s(j).x, s(j).x_dot, s(j).x_ddot);
                end
                %Z(:,j) = Z(:,j)./sc;
            end
            Z = inv_sc*Z;
            
            C = this.Sigma_w*H;
            B = eye(n,n) / (H'*this.Sigma_w*H + Rn) * C';
            this.gmp.W = this.gmp.W + (Z - this.gmp.W*H)*B;
            
            if (this.recursive_up)
                this.Sigma_w = this.Sigma_w - C*B;
            end
            
%             D = 1e-1*eye(n_ker, n_ker) / n_ker;
%             inv_S = inv(this.Sigma_w);
%             inv_R = inv(Rn);
%             w0 = this.gmp.W';
%             w =  (inv_S + D + H*inv_R*H')\(inv_S*w0 + H*inv_R*Z');
%             this.gmp.W = w';
            
        end

        function plotWeightsCovariance(this, ax)
            
            if (nargin < 2)
                figure;
                ax = axes();
            end
            imshow(this.Sigma_w, [min(this.Sigma_w(:)) max(this.Sigma_w(:))], 'InitialMagnification', 2000, 'Parent',ax);
            
        end
        
    end
    
    methods (Access = protected)
       
        function clearBatch(this)
            
            this.batch_s = [];
            this.batch_Z = [];
            this.batch_type = [];
            this.batch_r_n = [];
            
        end
        
    end
    
    properties (Access = protected)

        gmp % n_DoF GMP
        
        Sigma_w % weights covariance for trajectory update
        rv % noise variance for trajectory update
        recursive_up
        sync_up
        
        batch_s
        batch_Z
        batch_type
        batch_r_n

    end
    
end
